x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -246,53 +104,112 @@ export class SearchSessionService implements ISessionService {
 
     return from(getSearchRequest()).pipe(
       switchMap((request) => strategy.search(request, options, searchDeps)),
-      tapFirst((response) => {
-        if (searchRequest.id || !options.sessionId || !response.id || options.isRestore) return;
+      tap((response) => {
+        if (!options.sessionId || !response.id || options.isRestore) return;
         this.trackId(searchRequest, response.id, options, deps);
       })
     );
   }
 
-  // TODO: Generate the `userId` from the realm type/realm name/username
+  private updateOrCreate = async (
+    sessionId: string,
+    attributes: Partial<SearchSessionSavedObjectAttributes>,
+    deps: SearchSessionDependencies,
+    retry: number = 1
+  ): Promise<
+    | SavedObjectsUpdateResponse<SearchSessionSavedObjectAttributes>
+    | SavedObject<SearchSessionSavedObjectAttributes>
+    | undefined
+  > => {
+    const retryOnConflict = async (e: any) => {
+      this.logger.debug(`Conflict error | ${sessionId}`);
+      // Randomize sleep to spread updates out in case of conflicts
+      await sleep(100 + Math.random() * 50);
```

Comment:
Hmm, in your testing were you seeing conflicts? I don't imagine we'll need to randomize it like this since the updates are partial updates and shouldn't conflict with each other.

Reply:
You're getting conflicts when you're creating a session.
Assume 10 searches are starting. They try to update a non existing object (10/10 fail), then they immediately try to create the object (9/10 fail), then they all update the object (9/9 succeed).

We really need to use `upset` for this, but since we don't have that available to us, this is the most reliable mechanism I could come up with.

Reply:
I guess I don't understand how the randomization helps. With or without the randomization, the scenario you explained happens (10/10 fail updating, 9/10 fail creating, then 9/9 succeed after a timeout). Isn't that the case? Or does the randomization actually do something to prevent the 9 updates from conflicting?